src/axom/quest/MeshTester.hpp

```diff
@@ -33,12 +58,296 @@ enum class WatertightStatus : signed char
   CHECK_FAILED       ///< Calculation failed (possibly a non-manifold mesh)
 };
 
+inline detail::Triangle3 getMeshTriangle(axom::IndexType i, 
+                                         detail::UMesh* surface_mesh)
+{
+  SLIC_ASSERT( surface_mesh->getNumberOfCellNodes( i ) == 3);
+
+  detail::Triangle3 tri;
+
+  const axom::IndexType* triCell = surface_mesh->getCellNodeIDs( i );
+
+  const double* x = surface_mesh->getCoordinateArray( mint::X_COORDINATE );
+  const double* y = surface_mesh->getCoordinateArray( mint::Y_COORDINATE );
+  const double* z = surface_mesh->getCoordinateArray( mint::Z_COORDINATE );
+
+  for ( int n=0 ; n < 3 ; ++n )
+  {
+    const axom::IndexType nodeIdx = triCell[ n ];
+    tri[ n ][ 0 ] = x[ nodeIdx ];
+    tri[ n ][ 1 ] = y[ nodeIdx ];
+    tri[ n ][ 2 ] = z[ nodeIdx ];
+  }
+
+  return tri;
+}
 
 /// \name Mesh test and repair
 /// @{
 
 /*!
- * \brief Find self-intersections and degenerate triangles in a surface mesh.
+ * \brief Find self-intersections and degenerate triangles in a surface mesh
+ *  utilizing a Bounding Volume Hierarchy.
+ *
+ * \param [in] surface_mesh A triangle mesh in three dimensions
+ * \param [out] intersection Pairs of indices of intersecting mesh triangles
+ * \param [out] degenerateIndices indices of degenerate mesh triangles
+ * \param [in] intersectionThreshold Tolerance threshold for triangle 
+ * intersection tests (default: 1E-8)
+ * After running this function over a surface mesh, intersection will be filled
+ * with pairs of indices of intersecting triangles and degenerateIndices will
+ * be filled with the indices of the degenerate triangles in the mesh.
+ * Triangles that share vertex pairs (adjacent triangles in a watertight
+ * surface mesh) are not reported as intersecting.  Degenerate triangles
+ * are not reported as intersecting other triangles.
+ *
+ */
+#if defined(AXOM_USE_RAJA)
+template < typename ExecSpace, typename FloatType >
+void findTriMeshIntersectionsBVH(
+  mint::UnstructuredMesh< mint::SINGLE_SHAPE >* surface_mesh,
+  std::vector<std::pair<int, int> > & intersections,
+  std::vector<int> & degenerateIndices,
+  double intersectionThreshold = 1E-8)
+{
+  AXOM_PERF_MARK_FUNCTION( "findTriMeshIntersectionsBVH" );
+
+  SLIC_INFO("Running BVH intersection algorithm "
+            << " in execution Space: "
+            << axom::execution_space< ExecSpace >::name());
+
+  constexpr size_t POOL_SIZE   = (1024 * 1024 * 1024) + 1; 
+  umpire::ResourceManager& rm  = umpire::ResourceManager::getInstance();
+
+  // Use device pool for CUDA policy, host pool otherwise
+  std::string exec_name (axom::execution_space< ExecSpace >::name());
+  umpire::Allocator allocator =
+    (exec_name.find("[CUDA_EXEC]") != std::string::npos ? 
```

Comment:
Hmm...This can be made cleaner if we'd expand `execution_space` traits class and add another `constexpr bool` method. 

Specifically, I suggest we change the definition in `axom/core/execution_space.hpp`: https://github.com/LLNL/axom/blob/bac157c8834ac6b2e027701461ad5cfb50f0e9ab/src/axom/core/execution/execution_space.hpp#L75-L88

To something like this:
```
template < typename ExecSpace >
struct execution_space
{
  using loop_policy   = void;

  using reduce_policy = void;
  using atomic_policy = void;
  using sync_policy   = void;

  static constexpr bool async() noexcept { return false; };
  static constexpr bool valid() noexcept { return false; };
```
  **``static constexpr bool onDevice() noexcept {return false; };``**
```
  static constexpr char* name() noexcept { return (char*)"[UNDEFINED]"; };
  static int allocatorID() noexcept { return axom::INVALID_ALLOCATOR_ID; };
};
```
Subsequently, each template specialization of the ``execution_space`` traits class implements the ``onDevice`` method accordingly.

This would allow us to call ``axom::execution_space< ExecSpace >::onDevice()`` instead of relying on the name, which is something that would be generally useful as well.

Reply:
Added!